# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/Loader.cpp
# Commit: b59dcea26b32
# Full Hash: b59dcea26b327513f9843bd339ecfe4bb073e267
# Author: Tooru Fujisawa <arai_a@mac.com>
# Date: 2024-10-02 09:50:09
# Regressor Bug: 1899734
# File Overlap Count: 1
# Description:
#   Bug 1899734 - Part 7: Carry SubResourceNetworkMetadataHolder with SheetLoadData for complete and incomplete load, and propagate through dependencies. r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223353
# ==============================================================================

diff -r a15f046e5cdf -r b59dcea26b32 layout/style/Loader.cpp
--- a/layout/style/Loader.cpp	Tue Oct 01 12:17:22 2024 +0000
+++ b/layout/style/Loader.cpp	Tue Oct 01 12:17:23 2024 +0000
@@ -276,7 +276,8 @@
     IsAlternate aIsAlternate, MediaMatched aMediaMatches,
     StylePreloadKind aPreloadKind, nsICSSLoaderObserver* aObserver,
     nsIPrincipal* aTriggeringPrincipal, nsIReferrerInfo* aReferrerInfo,
-    const nsAString& aNonce, FetchPriority aFetchPriority)
+    const nsAString& aNonce, FetchPriority aFetchPriority,
+    already_AddRefed<SubResourceNetworkMetadataHolder>&& aNetworkMetadata)
     : mLoader(aLoader),
       mTitle(aTitle),
       mEncoding(nullptr),
@@ -309,18 +310,19 @@
       mCompatMode(aLoader->CompatMode(aPreloadKind)),
       mRecordErrors(
           aLoader && aLoader->GetDocument() &&
-          css::ErrorReporter::ShouldReportErrors(*aLoader->GetDocument())) {
+          css::ErrorReporter::ShouldReportErrors(*aLoader->GetDocument())),
+      mNetworkMetadata(std::move(aNetworkMetadata)) {
   MOZ_ASSERT(!aOwningNode || dom::LinkStyle::FromNode(*aOwningNode),
              "Must implement LinkStyle");
   MOZ_ASSERT(mTriggeringPrincipal);
   MOZ_ASSERT(mLoader, "Must have a loader!");
 }
 
-SheetLoadData::SheetLoadData(css::Loader* aLoader, nsIURI* aURI,
-                             StyleSheet* aSheet, SheetLoadData* aParentData,
-                             nsICSSLoaderObserver* aObserver,
-                             nsIPrincipal* aTriggeringPrincipal,
-                             nsIReferrerInfo* aReferrerInfo)
+SheetLoadData::SheetLoadData(
+    css::Loader* aLoader, nsIURI* aURI, StyleSheet* aSheet,
+    SheetLoadData* aParentData, nsICSSLoaderObserver* aObserver,
+    nsIPrincipal* aTriggeringPrincipal, nsIReferrerInfo* aReferrerInfo,
+    already_AddRefed<SubResourceNetworkMetadataHolder>&& aNetworkMetadata)
     : mLoader(aLoader),
       mEncoding(nullptr),
       mURI(aURI),
@@ -354,7 +356,8 @@
       mCompatMode(aLoader->CompatMode(mPreloadKind)),
       mRecordErrors(
           aLoader && aLoader->GetDocument() &&
-          css::ErrorReporter::ShouldReportErrors(*aLoader->GetDocument())) {
+          css::ErrorReporter::ShouldReportErrors(*aLoader->GetDocument())),
+      mNetworkMetadata(std::move(aNetworkMetadata)) {
   MOZ_ASSERT(mLoader, "Must have a loader!");
   MOZ_ASSERT(mTriggeringPrincipal);
   MOZ_ASSERT(!mUseSystemPrincipal || mSyncLoad,
@@ -367,7 +370,8 @@
     UseSystemPrincipal aUseSystemPrincipal, StylePreloadKind aPreloadKind,
     const Encoding* aPreloadEncoding, nsICSSLoaderObserver* aObserver,
     nsIPrincipal* aTriggeringPrincipal, nsIReferrerInfo* aReferrerInfo,
-    const nsAString& aNonce, FetchPriority aFetchPriority)
+    const nsAString& aNonce, FetchPriority aFetchPriority,
+    already_AddRefed<SubResourceNetworkMetadataHolder>&& aNetworkMetadata)
     : mLoader(aLoader),
       mEncoding(nullptr),
       mURI(aURI),
@@ -400,7 +404,8 @@
       mCompatMode(aLoader->CompatMode(aPreloadKind)),
       mRecordErrors(
           aLoader && aLoader->GetDocument() &&
-          css::ErrorReporter::ShouldReportErrors(*aLoader->GetDocument())) {
+          css::ErrorReporter::ShouldReportErrors(*aLoader->GetDocument())),
+      mNetworkMetadata(std::move(aNetworkMetadata)) {
   MOZ_ASSERT(mTriggeringPrincipal);
   MOZ_ASSERT(mLoader, "Must have a loader!");
   MOZ_ASSERT(!mUseSystemPrincipal || mSyncLoad,
@@ -1006,12 +1011,14 @@
  * state of the sheet they are clones off; make sure to call PrepareSheet() on
  * the result of CreateSheet().
  */
-std::tuple<RefPtr<StyleSheet>, Loader::SheetState> Loader::CreateSheet(
-    nsIURI* aURI, nsIContent* aLinkingContent,
-    nsIPrincipal* aTriggeringPrincipal, css::SheetParsingMode aParsingMode,
-    CORSMode aCORSMode, const Encoding* aPreloadOrParentDataEncoding,
-    const nsAString& aIntegrity, bool aSyncLoad,
-    StylePreloadKind aPreloadKind) {
+std::tuple<RefPtr<StyleSheet>, Loader::SheetState,
+           RefPtr<SubResourceNetworkMetadataHolder>>
+Loader::CreateSheet(nsIURI* aURI, nsIContent* aLinkingContent,
+                    nsIPrincipal* aTriggeringPrincipal,
+                    css::SheetParsingMode aParsingMode, CORSMode aCORSMode,
+                    const Encoding* aPreloadOrParentDataEncoding,
+                    const nsAString& aIntegrity, bool aSyncLoad,
+                    StylePreloadKind aPreloadKind) {
   MOZ_ASSERT(aURI, "This path is not taken for inline stylesheets");
   LOG(("css::Loader::CreateSheet(%s)", aURI->GetSpecOrDefault().get()));
 
@@ -1038,8 +1045,10 @@
     if (cacheResult.mState != CachedSubResourceState::Miss) {
       SheetState sheetState = SheetState::Complete;
       RefPtr<StyleSheet> sheet;
+      RefPtr<SubResourceNetworkMetadataHolder> networkMetadata;
       if (cacheResult.mCompleteValue) {
         sheet = cacheResult.mCompleteValue->Clone(nullptr, nullptr);
+        networkMetadata = cacheResult.mNetworkMetadata;
         mDocument->SetDidHitCompleteSheetCache();
       } else {
         MOZ_ASSERT(cacheResult.mLoadingOrPendingValue);
@@ -1049,7 +1058,7 @@
                          : SheetState::Pending;
       }
       LOG(("  Hit cache with state: %s", gStateStrings[size_t(sheetState)]));
-      return {std::move(sheet), sheetState};
+      return {std::move(sheet), sheetState, std::move(networkMetadata)};
     }
   }
 
@@ -1063,7 +1072,7 @@
       ReferrerInfo::CreateForExternalCSSResources(sheet);
   sheet->SetReferrerInfo(referrerInfo);
   LOG(("  Needs parser"));
-  return {std::move(sheet), SheetState::NeedsParser};
+  return {std::move(sheet), SheetState::NeedsParser, nullptr};
 }
 
 static Loader::MediaMatched MediaListMatches(const MediaList* aMediaList,
@@ -1890,7 +1899,8 @@
   auto data = MakeRefPtr<SheetLoadData>(
       this, aInfo.mTitle, /* aURI = */ nullptr, sheet, SyncLoad::No,
       aInfo.mContent, isAlternate, matched, StylePreloadKind::None, aObserver,
-      principal, aInfo.mReferrerInfo, aInfo.mNonce, aInfo.mFetchPriority);
+      principal, aInfo.mReferrerInfo, aInfo.mNonce, aInfo.mFetchPriority,
+      nullptr);
   MOZ_ASSERT(data->GetRequestingNode() == aInfo.mContent);
   if (isSheetFromCache) {
     MOZ_ASSERT(sheet->IsComplete());
@@ -2006,8 +2016,8 @@
   // Check IsAlternateSheet now, since it can mutate our document and make
   // pending sheets go to the non-pending state.
   auto isAlternate = IsAlternateSheet(aInfo.mTitle, aInfo.mHasAlternateRel);
-  auto [sheet, state] = CreateSheet(aInfo, eAuthorSheetFeatures, syncLoad,
-                                    StylePreloadKind::None);
+  auto [sheet, state, networkMetadata] = CreateSheet(
+      aInfo, eAuthorSheetFeatures, syncLoad, StylePreloadKind::None);
 
   LOG(("  Sheet is alternate: %d", static_cast<int>(isAlternate)));
 
@@ -2025,7 +2035,8 @@
   auto data = MakeRefPtr<SheetLoadData>(
       this, aInfo.mTitle, aInfo.mURI, sheet, SyncLoad(syncLoad), aInfo.mContent,
       isAlternate, matched, StylePreloadKind::None, aObserver, principal,
-      aInfo.mReferrerInfo, aInfo.mNonce, aInfo.mFetchPriority);
+      aInfo.mReferrerInfo, aInfo.mNonce, aInfo.mFetchPriority,
+      networkMetadata.forget());
 
   MOZ_ASSERT(data->GetRequestingNode() == requestingNode);
 
@@ -2143,6 +2154,7 @@
   // Now that we know it's safe to load this (passes security check and not a
   // loop) do so.
   RefPtr<StyleSheet> sheet;
+  RefPtr<SubResourceNetworkMetadataHolder> networkMetadata;
   SheetState state;
   bool isReusableSheet = false;
   if (aReusableSheets && aReusableSheets->FindReusableStyleSheet(aURL, sheet)) {
@@ -2150,7 +2162,7 @@
     isReusableSheet = true;
   } else {
     // For now, use CORS_NONE for child sheets
-    std::tie(sheet, state) = CreateSheet(
+    std::tie(sheet, state, networkMetadata) = CreateSheet(
         aURL, nullptr, principal, aParentSheet.ParsingMode(), CORS_NONE,
         aParentData ? aParentData->mEncoding : nullptr,
         u""_ns,  // integrity is only checked on main sheet
@@ -2162,9 +2174,9 @@
   MOZ_ASSERT(sheet);
   InsertChildSheet(*sheet, aParentSheet);
 
-  auto data =
-      MakeRefPtr<SheetLoadData>(this, aURL, sheet, aParentData, observer,
-                                principal, aParentSheet.GetReferrerInfo());
+  auto data = MakeRefPtr<SheetLoadData>(
+      this, aURL, sheet, aParentData, observer, principal,
+      aParentSheet.GetReferrerInfo(), networkMetadata.forget());
   MOZ_ASSERT(data->GetRequestingNode() == requestingNode);
 
   MaybeNotifyPreloadUsed(*data);
@@ -2258,7 +2270,7 @@
   }
 
   bool syncLoad = !aObserver;
-  auto [sheet, state] =
+  auto [sheet, state, networkMetadata] =
       CreateSheet(aURL, nullptr, triggeringPrincipal, aParsingMode, aCORSMode,
                   aPreloadEncoding, aIntegrity, syncLoad, aPreloadKind);
 
@@ -2268,7 +2280,7 @@
   auto data = MakeRefPtr<SheetLoadData>(
       this, aURL, sheet, SyncLoad(syncLoad), aUseSystemPrincipal, aPreloadKind,
       aPreloadEncoding, aObserver, triggeringPrincipal, aReferrerInfo, aNonce,
-      aFetchPriority);
+      aFetchPriority, networkMetadata.forget());
   MOZ_ASSERT(data->GetRequestingNode() == mDocument);
   if (state == SheetState::Complete) {
     LOG(("  Sheet already complete"));